File: TopCoder/src/main/java/com/leetcode/MaximumProductSubArray.java
package com.leetcode;

//https://leetcode.com/problems/maximum-product-subarray/
public class MaximumProductSubArray {

    public int maxProduct(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }
        int minProduct = 1;
        int maxProduct = 1;
        int finalMaxProduct = Integer.MIN_VALUE;
        for (int i = 0 ; i < nums.length ; i++) {
            int currMinProduct = Math.min(nums[i], Math.min(minProduct * nums[i], maxProduct * nums[i]));
            int currMaxProduct = Math.max(nums[i], Math.max(minProduct * nums[i], maxProduct * nums[i]));

            finalMaxProduct = Math.max(finalMaxProduct, currMaxProduct);
            minProduct = currMinProduct;
            maxProduct = currMaxProduct;
        }
        return finalMaxProduct;
    }

    public static void main(String[] args) {
        int[] arr = {-2, -2, -1};
        int result = new MaximumProductSubArray().maxProduct(arr);
        System.out.println(result);
    }
}

File: TopCoder/python/arr/MaximumProductSubArray.py
import sys
from typing import List
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        product: int = 1
        max_product = -sys.maxsize -1
        min_product = 1
        for num in nums:
            product = product * num
            min_product = min_product * num
            if product > max_product:
                max_product = product
            if min_product > max_product:
                max_product = min_product
            if product <= 0:
                product = 1
            if min_product == 0:
                min_product = 1
        return max_product


if __name__ == "__main__":
    arr = [2, 0, 7]
    sol = Solution()
    result = sol.maxProduct(arr)
    print(result)